subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformer.java

```diff
@@ -205,6 +205,18 @@ public InjectionPoint(Class<? extends Annotation> annotation, Class<?> injectedT
                 this.valueToInject = valueToInject;
             }
 
+            public InjectionPoint(Class<? extends Annotation> annotation, Object valueToInject) {
+                this(annotation, determineTypeFromAnnotation(annotation), valueToInject);
+            }
+
+            private static Class<?> determineTypeFromAnnotation(Class<? extends Annotation> annotation) {
+                Class<?>[] supportedTypes = annotation.getAnnotation(InjectionPointQualifier.class).supportedTypes();
+                if (supportedTypes.length != 1) {
+                    throw new IllegalArgumentException("Cannot determine supported type for annotation " + annotation.getName());
+                }
+                return supportedTypes[0];
+            }
+
```

Comment:
Should we have validation for these guys somewhere?

Reply:
What do you mean by validation? That the annotated method/type is a supertype of the supported types?

I would say yes, that is part of https://github.com/gradle/build-cache/issues/1398 though.